File: src/MyServiceBus/DefaultConstructorConsumerFactory.cs
using System.Threading.Tasks;

namespace MyServiceBus;

public class DefaultConstructorConsumerFactory<TConsumer> : IConsumerFactory<TConsumer>
    where TConsumer : class, new()
{
    public Task Send<TMessage>(ConsumeContext<TMessage> context,
        IPipe<ConsumerConsumeContext<TConsumer, TMessage>> next) where TMessage : class
    {
        var consumer = new TConsumer();
        var consumerContext = new ConsumerConsumeContextImpl<TConsumer, TMessage>(consumer, context);
        return next.Send(consumerContext);
    }
}

File: src/Java/myservicebus/src/main/java/com/myservicebus/DefaultConstructorConsumerFactory.java
package com.myservicebus;

import java.util.concurrent.CompletableFuture;

/**
 * Creates consumers using their parameterless constructor.
 */
public class DefaultConstructorConsumerFactory implements ConsumerFactory {
    @Override
    public <TConsumer, T> CompletableFuture<Void> send(Class<TConsumer> consumerType,
            ConsumeContext<T> context,
            Pipe<ConsumerConsumeContext<TConsumer, T>> next) {
        try {
            TConsumer consumer = consumerType.getDeclaredConstructor().newInstance();
            ConsumerConsumeContext<TConsumer, T> consumerContext = new ConsumerConsumeContext<>(consumer, context);
            return next.send(consumerContext);
        } catch (Exception ex) {
            return CompletableFuture.failedFuture(ex);
        }
    }
}

